### DISCUSSION TITLE
Zero-length chunks: are they expected?

### DISCUSSION THREAD

**Discussion by michalc** [encode/httpx#1733] [Potential Issue] [answered]
The last chunk in the below is of length zero:

```python
with httpx.stream('GET', 'https://www.gyan.dev/ffmpeg/builds/packages/ffmpeg-4.4-essentials_build.zip') as r:
    for chunk in r.iter_bytes():
        print(len(chunk))
```
is this expected?

It also looks like `iter_raw` _doesn't_ have a zero length chunk at the end:
```python
with httpx.stream('GET', 'https://www.gyan.dev/ffmpeg/builds/packages/ffmpeg-4.4-essentials_build.zip') as r:
    for chunk in r.iter_raw():
        print(len(chunk))
```
since in this case there is no content-encoding, I guess I expect that `iter_bytes == iter_raw`

----

Having a bit of a nose around the code, suspecting it's from https://github.com/encode/httpx/blob/47d712c01cdde6574b2a94e43a94e5ae6e4f2faf/httpx/_models.py#L1483 where `decoder.flush()` returns an empty `bytes` instance, this is passed to `chunker.decode`, and since `chunk_size` is `None`, a list containing this empty `bytes` instance is returned https://github.com/encode/httpx/blob/47d712c01cdde6574b2a94e43a94e5ae6e4f2faf/httpx/_decoders.py#L172

and then each element of this list is yielded to the calling code

----

This is related to https://github.com/uktrade/stream-unzip/issues/13, where the code in stream-unzip does not expect any zero length chunks. Potentially stream-unzip should be robust against this case... but wasn't sure if this was desirable in httpx, so mentioning it here in case it is an issue to be addressed.

**Comment by lovelydinosaur** [comment, marked as answer]
Thanks @michalc - I've escalated this into #1739.

Short version is that we're not currently making any assertions one way or the other about if any chunks might be zero-length, but it'd clearly be preferable from a UX perspective if it *didn't* ever do this.
